lib/Dialect/Arc/ArcOps.cpp

```diff
@@ -307,6 +308,30 @@ LogicalResult ModelOp::verify() {
   return success();
 }
 
+LogicalResult ModelOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  if (!getInitialFn().has_value())
+    return success();
+
+  auto referencedOp =
+      symbolTable.lookupNearestSymbolFrom(*this, getInitialFnAttr());
+  if (!referencedOp)
+    return emitError("Cannot find declaration of initializer function '")
+           << *getInitialFn() << "'.";
+  auto funcOp = dyn_cast<func::FuncOp>(referencedOp);
+  if (!funcOp) {
+    auto diag = emitError("Referenced initializer must be a 'func.func' op.");
+    diag.attachNote(referencedOp->getLoc()) << "Initializer declared here:";
+    return diag;
+  }
+  if (!llvm::equal(funcOp.getArgumentTypes(), getBody().getArgumentTypes())) {
+    auto diag = emitError("Arguments of initializer function must match "
+                          "arguments of model body.");
+    diag.attachNote(referencedOp->getLoc()) << "Initializer declared here:";
+    return diag;
+  }
```

Comment:
Nice idea checking that the function signature is reasonable.

This makes me think that we might want to create something like a `arc.lowered_model` op in the future which has no body region, but has symbol refs to the init, eval, and final functions, and a symbol ref to the data layout (@maerhart has been working on a prototype to move the sim memory layout into a struct-like op). That would allow us to keep all information about the generated models around until the very end, without restricting the conversion to functions and further optimizations. Anyway, very much future stuff :wink: